typescript/sdk/src/ism/metadata/routing.ts

```diff
@@ -56,3 +57,49 @@ export class RoutingMetadataBuilder implements MetadataBuilder {
     };
   }
 }
+
+export class DefaultFallbackRoutingMetadataBuilder extends RoutingMetadataBuilder {
+  public async build(
+    context: MetadataContext<WithAddress<RoutingIsmConfig>>,
+    maxDepth = 10,
+  ): Promise<string> {
+    const originChain = this.baseMetadataBuilder.multiProvider.getChainName(
+      context.message.parsed.origin,
+    );
+
+    const isRouted = !!context.ism.domains[originChain];
+    if (isRouted) {
+      return super.build(context, maxDepth);
+    }
+
+    if (context.ism.type !== IsmType.FALLBACK_ROUTING) {
+      throw new Error(
+        `Origin domain ${originChain} is not enrolled in DomainRoutingIsm`,
+      );
+    }
+
+    const destinationProvider =
+      this.baseMetadataBuilder.multiProvider.getProvider(
+        context.message.parsed.destination,
+      );
+    const fallbackIsm = DefaultFallbackRoutingIsm__factory.connect(
+      context.ism.address,
+      destinationProvider,
+    );
+    const defaultIsmAddress = await fallbackIsm.module(
+      context.message.parsed.origin,
+    );
+
+    const ismReader = new EvmIsmReader(
+      this.baseMetadataBuilder.multiProvider,
+      context.message.parsed.destination,
+    );
+    const defaultIsmConfig = await ismReader.deriveIsmConfig(defaultIsmAddress);
```

Comment:
this is clever I guess but kinda awkward that we have the reader inside the metadata builder itself?
does the reader not support deriving the fallback tree already?

Reply:
We do not currently do that

Reply:
this is basically "lazy" ISM derivation
this requires a lot of network round trips so will contribute significantly to latency relaying messages with a default fallback ISM